File: generate-url.py
#!/usr/bin/env python
# Copyright (C) 2013 Luca Filipozzi <luca.filipozzi@gmail.com>
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.

import hmac
import base64

_host = "www.example.org"
_src  = "/foo"
_hash = "2816d3b56ebeaabd4af3a31d9b1c17f545a8898a"
_type = "application/x-gzip"
_file = "blah-1.2.tar.gz"

_key = "secret"
_msg = "/".join([_hash, base64.b16encode(_type).lower(), _file])

_hmac = hmac.new(_key, _msg).hexdigest()

_rest = "/".join([_src, _hmac, _msg])
print "https://" + _host + _rest

File: generate-url.lua
#!/usr/bin/env lua
-- Copyright (C) 2013 Luca Filipozzi <luca.filipozzi@gmail.com>
--
-- All rights reserved.
--
-- Redistribution and use in source and binary forms, with or without
-- modification, are permitted provided that the following conditions are met:
--
-- 1. Redistributions of source code must retain the above copyright notice,
--    this list of conditions and the following disclaimer.
--
-- 2. Redistributions in binary form must reproduce the above copyright notice,
--    this list of conditions and the following disclaimer in the documentation
--    and/or other materials provided with the distribution.
--
-- THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
-- ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
-- WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
-- DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
-- FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
-- DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
-- SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
-- CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
-- OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
-- OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

local md5 = require("md5")

function hmac_md5(_key, _msg)
  if type(_key) ~= "string" or type(_msg) ~= "string" then
    return nil
  end

  local _blocksize = 64 -- 512 bits for md5

  -- keys longer than blocksize are shortened
  if string.len(_key) > _blocksize then -- truncate
    _key = md5.sum(_key)
  end

  -- keys shorter than blocksize are zero-padded
  if string.len(_key) < _blocksize then
    _key = _key .. string.rep(string.char(0x00), _blocksize - string.len(_key))
  end

  local _opad = md5.exor(string.rep(string.char(0x5c), _blocksize), _key)
  local _ipad = md5.exor(string.rep(string.char(0x36), _blocksize), _key)

  return md5.sumhexa(_opad .. md5.sum(_ipad .. _msg))
end

function base16_encode(_str)
  return (_str:gsub('.', function(_c)
    return string.format('%02x', string.byte(_c))
  end))
end

local _host = "www.example.org"
local _src  = "/foo"
local _hash = "2816d3b56ebeaabd4af3a31d9b1c17f545a8898a"
local _type = "application/x-gzip"
local _file = "blah-1.2.tar.gz"

local _key = "secret"
local _msg = _hash .. "/" .. base16_encode(_type) .. "/" .. _file

local _hmac = hmac_md5(_key, _msg)

local _rest = _src .. "/" .. _hmac .. "/" .. _msg
print("https://" .. _host .. _rest)
